src/main/scala/org/ergoplatform/modifiers/history/popow/PoPowAlgos.scala

```diff
@@ -0,0 +1,270 @@
+      }
+
+    if (leftChain.headOption.exists(rightChain.headOption.contains(_))) Some(leftChain(lcaIndex(1))) else None
+  }
+
+  /**
+    * Computes NiPoPow proof for the given `chain` according to a given `params`.
+    */
+  def prove(chain: Seq[PoPowHeader])(params: PoPowParams): PoPowProof = {
+    require(chain.lengthCompare(params.k + params.m) >= 0,
+      s"Can not prove chain of size < ${params.k + params.m}")
+    require(chain.head.header.isGenesis, "Can not prove non-anchored chain")
+
+    @scala.annotation.tailrec
+    def provePrefix(anchoringPoint: PoPowHeader,
+                    level: Int,
+                    acc: Seq[PoPowHeader] = Seq.empty): Seq[PoPowHeader] =
+      if (level >= 0) {
+        val subChain = chain.dropRight(params.k)
+          .filter(h => maxLevelOf(h.header) >= level && h.height >= anchoringPoint.height) // C[:−k]{B:}↑µ
+        val goodChain = goodSuperChain(chain.map(_.header), subChain.map(_.header), level)(params)
+        if (subChain.size >= params.m && goodChain) {
+          provePrefix(subChain(subChain.size - params.m), level - 1, acc ++ subChain)
+        } else {
+          provePrefix(anchoringPoint, level - 1, acc ++ subChain)
+        }
+      } else {
+        acc
+      }
+
+    val suffix = chain.takeRight(params.k)
+    val maxLevel = suffix.head.interlinks.size - 1
+    val prefix = provePrefix(chain.head, maxLevel).distinct.sortBy(_.height)
+    PoPowProof(params.m, params.k, prefix, suffix)
+  }
+
+  /**
+    * "Goodness" bounds the deviation of superblocks of a certain level from their expected mean.
+    */
+  def goodSuperChain(chain: Seq[Header], superChain: Seq[Header], level: Int)(params: PoPowParams): Boolean =
+    superChainQuality(chain, superChain, level)(params) && multiLevelQuality(chain, superChain, level)(params)
+
+  /**
+    * Checks μ-local goodness of μ-superchain of size `superChainSize` with respect to the underlying
+    * chain of size `underlyingChainSize` and security parameter `d`.
+    */
+  def locallyGood(superChainSize: Int, underlyingChainSize: Int, level: Int, d: Double): Boolean =
+    superChainSize > ((1 - d) * math.pow(2, -level) * underlyingChainSize)
+
+  /**
+    * Checks whether a given `superChain` is qualified against a given full `chain` at a given `level`.
+    *
+    * @param chain      - Full chain (C)
+    * @param superChain - Super-chain of level µ (C↑µ)
+    * @param level      - Level of super-chain (µ)
+    */
+  def superChainQuality(chain: Seq[Header], superChain: Seq[Header], level: Int)
+                       (params: PoPowParams): Boolean = {
+    val downChain = chain
+      .filter(h => h.height >= superChain.head.height && h.height <= superChain.last.height) // C[C↑µ[0]:C↑µ[−1]] or C'↓
+    @scala.annotation.tailrec
+    def checkLocalGoodnessAt(mValue: Int): Boolean = {
+      val superChainSuffixSize = superChain.takeRight(mValue).size
+      val downChainSuffixSize = downChain.takeRight(mValue).size
```

Comment:
Superchain quality states that for all m' ≥ m, it holds that 
`local-good δ (C↑ μ [−m' :], C↑ μ [−m' :]↓ , μ)`
As far as i can tell this operator means
`(C↑)↓ μ [−m' :]`
and this is different from the required
`C↑ μ [−m' :]↓`
Is it really correct?

Reply:
Right, seems we need to take M elements from super-chain first and then take the size of its projection on the down-chain

Reply:
@oskin1 please make an issue and a fix based on #990

Reply:
@ivanovs-4 @oskin1 please review my fix

Reply:
It looks like that now it conforms to the specification.